Framework/API/src/ExperimentInfo.cpp

```diff
@@ -219,7 +217,8 @@ void ExperimentInfo::setInstrument(const Instrument_const_sptr &instr) {
   // instrument may now suddenly be valid, so we have to reinitialize the
   // detector grouping. Also the index corresponding to specific IDs may have
   // changed.
-  invalidateSpectrumDefinitions();
+  std::fill(m_spectrumDefinitionNeedsUpdate.begin(),
```

Comment:
Yes, it is basically a problem of `MatrixWorkspace` which is initialized in many steps (some of which are optional):

- (copy) constructor
- `MatrixWorkspace::initialize`
- `ExperimentInfo::setInstrument` or `WorkspaceFactory::initializeFromParent`
- setting detector IDs for spectra